### ISSUE TITLE
Create docker lumberjack driver

### ISSUE THREAD

**Issue by andrewvc** [elastic/libbeat#37]
Moved from (https://github.com/elastic/logstash/issues/3646)

After filebeat is ready we should create a docker driver based on it. After discussing this issue with @jordansissel we came to the conclusion that the driver should act as follows (apologies if my memory has any errors @jordansissel ). 

1. It should buffer log messages to disk in some sort of circular buffer
2. It should send those log messages to the remote server using lumberjack
3. After receiving an application-level acknowledgement (hereafter 'ack') it may free that space on the disk
4. It may replay messages in the absence of an ack

The acks have relevance for [persistence](https://github.com/elastic/logstash/issues/3693)

For examples of existing docker drivers see [this directory](https://github.com/docker/docker/tree/master/daemon/logger) in the docker source code. The driver will need to be submitted to docker/docker as a pull request.

**Comment by kimchy** [comment]
I think the title is a bit misleading. It is not a docker lumberjack driver, to me, it is more more of a docker beat. Lumberjack is just the way a beat can talk to an LS instance, and it also supports talking to ES directly. So, any docker specific behavior needs to exist in the Docker beat, not in lumberjack.

**Comment by andrewvc** [comment]
@kimchy I'm fine changing the terminology, but if we do we should change the protocol name to 'filebeat' or something from lumberjack. 

The reason here being that Docker does not support plugins because go does not support dynamic linking. Hence, we'll need to have our code submitted to docker as a PR to be merged with the main Docker tree (like all the other log drivers). At that point there's no 'beat', really, just a piece of our go code in the docker source tree that happens to be identical to what's in filebeat. You can't install it separately, you must download the latest docker with that source code directly compiled into it. 

I think that makes it hard to brand beyond the protocol name. That's why the Graylog and FluentD plugins are labeled GELF (the graylog protocol) and Fluent (the fluentd protocol) respectively.

**Comment by andrewvc** [comment]
@kimchy WRT to talking to ES directly I think that should be a separate logging driver since the protocol is completely different (though they would share code I imagine) and the config options would be very different as well. At present the docker logging driver names map 1:1 with the protocols they work with. While you could have a multi-protocol driver via parameters I think that may be confusing.

I imagine the UX being something like

`docker run --log-driver=lumberjack --log-opt url=http://localhost:9200 --log-opt index_name=mylogs ...`
and
`docker run --log-driver=elasticsearch --log-opt host=localhost --log-opt port=123 --log-opt tag=mymachine ...`

**Comment by tsg** [comment]
Regarding the persistency aspect, I was thinking Beats in general would support that by chaining them. For example, one could run Packetbeat with the file output to write to rotating files and then Filebeat/LSF to send them with the "at least once" guarantee. This feels to me cleaner then maintaining the circular buffer as a libbeat internal.

But I see how this doesn't quite work for Docker because it would require a separate process on the host,  which I guess it's what the docker plugins are trying to avoid (one can run LSF on `/var/lib/docker/container/...` right now).

**Comment by andrewvc** [comment]
@tsg would that be possible with multiple threads / goroutines? One 'persistor' routine and one 'network' routine? 

Apologies for using the term circular buffer, I didn't mean that literally. I think rotating log files is a great way to do that.

**Comment by tsg** [comment]
@andrewvc yes, that should be doable, just more difficult because it would mean moving almost all of LSF into libbeat. Not necessarily a bad thing, then all beats would benefit from persistency at the switch of a configuration option.

**Comment by andrewvc** [comment]
@tsg I'm not a huge fan of the Docker plugin model TBH, but I'm glad to hear its workable and has some side benefits :).

**Comment by marceldegraaf** [comment]
@andrewvc @tsg is someone already working on this integration? If not, I'd be interested to pick it up and submit a PR for it.

**Comment by ruflin** [comment]
@marceldegraaf not yet.

**Comment by marceldegraaf** [comment]
Have you considered doing something like (or integrating with) [logspout](https://github.com/gliderlabs/logspout)?

What that does is run in a separate Docker container on a host, which listens to log output from all other containers on the host. It then routes log lines to one or more endpoints, including a third party [Logstash UDP module](https://github.com/looplab/logspout-logstash) to forward log lines to Logstash over UDP.

Dockerbeat could work in a similar way: listen for log lines of all containers running on a host, and forward them to Logstash, or to Elasticsearch directly, using `libbeat` and the improved `lumberjack` protocol.

**Comment by marceldegraaf** [comment]
Oh, I see that Bol.com also created a [Redis/Logstash plugin](https://github.com/bolcom/logspout-redis-logstash) for Logspout.

**Comment by urso** [comment]
Let's add a feature request to: https://github.com/Ingensi/dockerbeat

**Comment by mrkschan** [comment]
Hey, just my 2cents, I'm not sure if we really need a docker log driver that would handle log persistence.

At the current moment,

* fluent log driver buffers message in memory and retry later if there's error. [1]
* gelf log driver just report error. [2]
* aws log driver just report unrecoverable error. [3]

As long as the error mode is well documented, the users would know how to configure their setup according to their need.

BTW, one simple ELK solution is: docker -> syslog -> filebeat -> Logstash / ElasticSearch and that would already handles persistence well.

[1]: https://github.com/docker/docker/blob/e3c472426ff1dccf083e41dbbfdf5935921bf330/daemon/logger/fluentd/fluentd.go#L71
[2]: https://github.com/docker/docker/blob/e3c472426ff1dccf083e41dbbfdf5935921bf330/daemon/logger/gelf/gelf.go#L98
[3]: https://github.com/docker/docker/blob/32964def832f76f29991100c76af09513de19693/daemon/logger/awslogs/cloudwatchlogs.go#L218

**Comment by andrewvc** [comment]
@mrkschan just because these drivers act that way doesn't mean its a good idea. Logs have to go *somewhere*, and memory is a dangerous somewhere because it can fill up easily if the remote server is down and crash the host.

The point of buffering to disk is to prevent the log service (logstash in this case), from affecting website uptime. Logstash going down shouldn't take down docker containers logging to it.

While disk can fill up, there's a much larger safety buffer there.

**Comment by mrkschan** [comment]
@andrewvc I agree with that and so docker -> syslog -> logstash forwarder is used in my company.

And, I think it's better to have an option for the user to choose if logs are persisted by docker logging driver before sending out. Why don't just start simple and introduce a non persisted one first?

**Comment by jordansissel** [comment]
>  I'm not sure if we really need a docker log driver that would handle log persistence

There's lots of options available. Logstash supports both fluentd and gelf protocols, so in practice, if someone can accept lossy log transmission, then there are two existing docker log drivers that can do this.

**Comment by jordansissel** [comment]
> I think it's better to have an option for the user to choose if logs are persisted

That choice is already available today by choosing fluentd or gelf instead of beats.

**Comment by mrkschan** [comment]
@jordansissel yea, I agree that users can do some google search and make a setup to use ELK. But, can they simply read the docker docs to do so?

**Comment by andrewvc** [comment]
@mrkschan that's a good point. I do think that it would be more discoverable for us to have a simple logstash driver that worked the same way as the GELF and and fluentd drivers. 

However, there's a real cost there when people set that up, then their whole production cluster goes down because they filled up their in memory buffer. Ultimately logstash will be blamed.

I think this is a case where logstash may get hurt a little bit by not having a new shiny docker driver (and I really want us to have one ASAP!). But IMHO the GELF and FluentD drivers are not production ready as is. IMHO we should ship something that operators can depend on, even if it means a delay.

**Comment by mrkschan** [comment]
IIRC, fluent log driver keep at most N events in memory. N can be configured. After the buffer is full, fluent driver reports error.

**Comment by urso** [comment]
I think the logspout approach sounds much better then the native docker-driver. Only advantage of docker-driver is... marketing... Approach by logspout works by using docker client to connect to docker host and [forward all logs](https://github.com/gliderlabs/logspout/blob/master/router/pump.go#L146) (see [api doc](https://godoc.org/github.com/fsouza/go-dockerclient#Client.Logs)). This approach supports collecting logs from remote docker hosts if [json-file](https://docs.docker.com/reference/logging/overview/) driver is used (the default one). This approach can easily be added to [dockerbeat](https://github.com/Ingensi/dockerbeat) already using the docker client approach to collect some container metrics.

Plus when relying on json-file (default) log driver and docker and connecting using the client api, users can just get logs out of the box without configuring any log-driver. Just attach to running host. The json-file log-driver also supports log rotation. Plus dockerbeat gets container meta data, cpu stats, network stats, memory stats and io stats. Think monitoring docker hosts and correlating logs to other metrics from within kibana. There is more potential then just getting some logs + you one can add monitoring to an already running docker installation almost on the fly. Think monitoring a complete docker-swarm, without even configuring anything on the docker site. How awesome is this? Being based on libbeat it can publish data to logstash for further processing or just index into elasticsearch.

Also keep in mind beat->logstash integration yet being in its infancy (was just added support in beta4). There might be changes to the protocol we would have a hard time to do if we would integrate libbeat into docker right now in order to support whatever versions people might want to use. Once docker environment is running, don't expect people to upgrade it with every docker version released => Big PITA for beats developers and support guys.

**Comment by mrkschan** [comment]
@urso you're right, docker upgrade is PITA. I did that twice in a production environment. With the logspout approach, my only concern is if that will create duplicate set of log lines. One from json-file logging driver and one from dockerbeat for log persistence.

**Comment by urso** [comment]
I'd assume the logspout approach will mostly work like filebeat: tail logs only + continue where we left since restart ([log api has 'since' options](https://docs.docker.com/reference/api/docker_remote_api_v1.20/#get-container-logs)). The json-file driver will do keep messages on docker host disk (think buffering). One should configure  --log-opt max-size=... though.

In conjunction to dockerbeat pulling logs, one can use the docker syslog driver to push logs. Like logstash already having syslog support. Or use this guys approach  [basically hacking a syslogbeat based on filebeat stdin support](https://discuss.elastic.co/t/high-rate-remote-syslog-into-filebeat/33092)

So in the end you can opt between pull based (dockerbeat) with buffering (relying on docker default log handling) or push based (syslog) processing and have your logs index into Elasticsearch.

If you insist on driver, why no driver starting an extra process and forward all logs to this process it's stdin ? This will work with filebeat and we can progress without having to deal with docker release cycles.

**Comment by mrkschan** [comment]
I do agree that progressing with docker release cycle may be an issue. Though, correct me if I'm wrong, the proposed logspout approach is like filebeat. So, why don't just use filebeat instead?

**Comment by urso** [comment]
filebeat approach is already possible. On my machine docker stores log files into /var/lib/docker/containers/<id>/<id>-json.log . But filebeat needs to be run on docker-host, plus Logstash is required to process the json content in these log files (say logstash json filter must be used). The logspout approach on the other hand collects logs from (potentially multiple remote) docker hosts and can directly index data right into Elasticsearch. From my point of view the advantage of dockerbeat is less setup/configuration overhead + flexibility in installation and scaling.

**Comment by mrkschan** [comment]
That sounds great! The persistence aspect is handled via json-file log driver while dockerbeat pull logs from a cluster and forward to ES / ELK. :heart:

**Comment by mrkschan** [comment]
FYI, docker is moving towards a non-embedded plugin approach via Rest API for plugin registration [1]. You may take a look at how to attach a Volumn plugin from their docs [2].

Given these info, it's not a good timing to develop a embedded logging driver.

[1]: http://docs.docker.com/engine/extend/plugin_api/
[2]: http://docs.docker.com/engine/extend/plugins_volume/

**Comment by mrkschan** [comment]
fyi, an update from docker. https://github.com/docker/docker/issues/14949#issuecomment-171973008

In short, we may have a "beat" that subscribes to a UDP/TCP/unix-socket log stream directly from the docker daemon.
